src/multiversioning.zig

```diff
@@ -0,0 +1,892 @@
+            release_triple.major,
+            release_triple.minor,
+            release_triple.patch,
+        });
+    }
+
+    pub fn max(a: Release, b: Release) Release {
+        if (a.value > b.value) {
+            return a;
+        } else {
+            return b;
+        }
+    }
+};
+
+pub const ReleaseTriple = extern struct {
+    patch: u8,
+    minor: u8,
+    major: u16,
+
+    comptime {
+        assert(@sizeOf(ReleaseTriple) == 4);
+        assert(stdx.no_padding(ReleaseTriple));
+    }
+
+    pub fn parse(string: []const u8) error{InvalidRelease}!ReleaseTriple {
+        var parts = std.mem.splitScalar(u8, string, '.');
+        const major = parts.first();
+        const minor = parts.next() orelse return error.InvalidRelease;
+        const patch = parts.next() orelse return error.InvalidRelease;
+        if (parts.next() != null) return error.InvalidRelease;
+        return .{
+            .major = std.fmt.parseUnsigned(u16, major, 10) catch return error.InvalidRelease,
+            .minor = std.fmt.parseUnsigned(u8, minor, 10) catch return error.InvalidRelease,
+            .patch = std.fmt.parseUnsigned(u8, patch, 10) catch return error.InvalidRelease,
+        };
+    }
+};
+
+test "ReleaseTriple.parse" {
+    const tests = [_]struct {
+        string: []const u8,
+        result: error{InvalidRelease}!ReleaseTriple,
+    }{
+        // Valid:
+        .{ .string = "0.0.1", .result = .{ .major = 0, .minor = 0, .patch = 1 } },
+        .{ .string = "0.1.0", .result = .{ .major = 0, .minor = 1, .patch = 0 } },
+        .{ .string = "1.0.0", .result = .{ .major = 1, .minor = 0, .patch = 0 } },
+
+        // Invalid characters:
+        .{ .string = "v0.0.1", .result = error.InvalidRelease },
+        .{ .string = "0.0.1v", .result = error.InvalidRelease },
+        // Invalid separators:
+        .{ .string = "0.0.0.1", .result = error.InvalidRelease },
+        .{ .string = "0..0.1", .result = error.InvalidRelease },
+        // Overflow (and near-overflow):
+        .{ .string = "0.0.255", .result = .{ .major = 0, .minor = 0, .patch = 255 } },
+        .{ .string = "0.0.256", .result = error.InvalidRelease },
+        .{ .string = "0.255.0", .result = .{ .major = 0, .minor = 255, .patch = 0 } },
+        .{ .string = "0.256.0", .result = error.InvalidRelease },
+        .{ .string = "65535.0.0", .result = .{ .major = 65535, .minor = 0, .patch = 0 } },
+        .{ .string = "65536.0.0", .result = error.InvalidRelease },
+    };
+    for (tests) |t| {
+        try std.testing.expectEqualDeep(ReleaseTriple.parse(t.string), t.result);
+    }
+}
+
+pub const MultiversionHeader = extern struct {
+    // When slicing into the binary:
+    // checksum(section[past_offset..past_offset+past_size]) == past_checksum.
+    // This is then validated when the binary is written to a memfd or similar.
+    // TODO: Might be nicer as an AoS? It's control plane state.
+    pub const PastReleases = extern struct {
```

Comment:
For each release, could we also include:

- `git_commit` (this is part of each binary anyway, but would be convenient here for introspection/logging)
- Maybe a boolean (possibly within a `u64` flags bitset, for future use and `visits`?) of whether or not the binary has debug info?
- `release_client_min` could be useful too.